[FSDP] Refactor casting of grad to full param dtype

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#81574 [FSDP] Refactor casting of grad to full param dtype**

I noticed a comment was repeated, so I wanted to refactor the duplication. Refer to the comment for the explanation.

Commit:
[FSDP] Refactor casting of grad to full param dtype

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2949,20 +2949,7 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
-                    ):
-                        # Cast gradients back to the full parameter precision so that
-                        # optimizer.step() happens in full precision.
-                        orig_param_grad_data = output
```
Some subtle things happening here:
1. If we compare this block with the block below, we notice that `orig_param_grad_data` here is set to `output` and not `output.data`. I wonder if this is a bug.
Once we set `output.data = output.data.to(dtype=param.data.dtype)` in the next line, the original storage of `output` loses its reference and can be garbage collected.
The third line `orig_param_grad_data.record_stream(...)` may be a no-op because `orig_param_grad_data`'s storage has been set to the new full parameter precision storage (only, the data transfer may not have completed). In that case, the `record_stream()` is incorrectly called for the new storage instead of the old storage. The old storage's memory may be reused early before the transfer completes.
2. I think that `param.data.dtype` is always the same as `param.dtype`. In my understanding, `.data` returns a new Python `Tensor` object with a new storage object but the same values of storage attributes. I am not sure what the `.data` call here is meant to do. If it is needed, I can add it back.

awgu commented:
@pytorchbot rebase

Commit:
Update on "[FSDP] Refactor casting of grad to full param dtype"


I noticed a comment was repeated, so I wanted to refactor the duplication. Refer to the comment for the explanation.

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3137,6 +3111,30 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
+            or self._mixed_precision_enabled_for_reduce()
+        ):
+            low_prec_grad_data = grad.data
+            grad.data = grad.data.to(dtype=full_param_dtype)
```
`non_blocking=True`?

awgu replied:
`non_blocking=True` seems safe here, but I am also not entirely sure. The [discussion online](https://discuss.pytorch.org/t/should-we-set-non-blocking-to-true/38234/17?u=agu) mostly refers to CPU <-> GPU transfers, not GPU <-> GPU transfers with different `dtype`. I am going to leave it as `non_blocking=False` (default) since I am not fully confident it will be safe.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3137,6 +3111,30 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
+        full_param_dtype: torch.dtype,
+    ):
+        """
+        Casts gradients back to the full parameter dtype so that the optimizer
```
nit: Casts gradient?

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3137,6 +3111,30 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
+    def _cast_grad_to_full_param_dtype(
+        self,
+        grad: torch.Tensor,
+        full_param_dtype: torch.dtype,
```
shall we pass the param, or just make this a generalized dtype?

rohan-varma reviewed (commented):
LGTM

Commit:
Update on "[FSDP] Refactor casting of grad to full param dtype"


I noticed a comment was repeated, so I wanted to refactor the duplication. Refer to the comment for the explanation.

[ghstack-poisoned]

awgu commented:
@pytorchbot rebase

Commit:
Update on "[FSDP] Refactor casting of grad to full param dtype"


I noticed a comment was repeated, so I wanted to refactor the duplication. Refer to the comment for the explanation.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Refactor casting of grad to full param dtype"


I noticed a comment was repeated, so I wanted to refactor the duplication. Refer to the comment for the explanation.

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3157,6 +3126,35 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
+        """
+        self._assert_state(TrainingState_.BACKWARD_POST)
+        if (
+            not self._low_precision_hook_enabled()
```
```
def self._low_precision_hook_enabled(self):
    return (
        self.communication_hook is not None
        and self.communication_hook in LOW_PRECISION_HOOKS
    )
```
For the strategies that are not `NO_SHARD`, `self.communication_hook is not None` is `False`, so `not self._low_precision_hook_enabled()` is always `True`. This means that for the non-`NO_SHARD` strategies, the semantics of this method are preserved.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3117,25 +3104,7 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
-                    # would thus be in this reduced precision, or
-                    # 2) parameters did not have precision reduced, but grads
-                    # had reduced precision for communication.
-                    # If a lower precision hook is registered, gradients are casted
-                    # back by the hook.
-                    # However, if a low precision hook is attached to the model.
-                    # casting happens inside the hook.
```
Seems like a duplicate comment? cc: @aovladi

awgu commented:
Rebased on top of https://github.com/pytorch/pytorch/pull/81711/

Commit:
Update on "[FSDP] Refactor casting of grad to full param dtype"


I noticed a comment was repeated, so I wanted to refactor the duplication. Refer to the comment for the explanation.

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3157,6 +3126,35 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
+        gradients would be in that reduced precision, or
+        2. parameters were not in reduced precision but gradients were in
+        reduced precision for communication.
+        However, if a low precision communication hook is registered, then this
```
This logic is getting a bit complicated - maybe it would be good to figure out a way to simplify this, such as make the hooks composable and register a low precision comm hook that does the casting if we end up doing mixed precision for gradient communication.

rohan-varma reviewed (approved):
One small comment for future follow-up work. Otherwise LGTM

awgu commented:
@pytorchbot merge
